homeassistant/components/energy/websocket_api.py

```diff
@@ -294,12 +294,11 @@ def _combine_sum_statistics(
         return {key: result[key] for key in sorted(result)}
 
     def _calculate_deltas(sums: dict[float, float]) -> dict[float, float]:
-        prev: float | None = None
+        prev: float = 0.0
         result: dict[float, float] = {}
         for period, sum_ in sums.items():
-            if prev is not None:
-                result[period] = sum_ - prev
-            prev = sum_
+            result[period] = sum_ - prev
+            prev = sum_ if sum_ is not None else 0.0
         return result
```

Comment:
Obviously I am not familiar with this codebase, but by the variable names, it sounds like this function is being passed the summed values for each sub-period within the requested period of time, which would suggest that defaulting to 0 would only work if the usage prior to the period being calculated is in fact 0. That is, would this show and calculate based on 10001 kWh or based on 1 kWh if the summed value of the grid consumption sensor/statistics prior to the day is some arbitrary high number, let's say 10000 kWh, and the first hour of the day uses 1 kWh? Your example screenshots and the tests both don't appear to have anything in place for this (in both cases, the values truly do start at 0), but it could be that there is some normalization that happens first in regular usage that I haven't been able to find.

Reply:
Yes, you are right, this function gets the summed value of each hour within the desired time period (fetched from statistics and summed up by `_combine_sum_statistics`).

Suppose we passed the following to `_calculate_deltas`:

12:00:00 -> 300
13:00:00 -> 400
14:00:00 -> 500

The value for "12:00:00" is dropped because the initial value of "prev" is always "None" when calling the function and thus the value is skipped in L300. In this case we just calculate the delta for 13:00:00 - 14:00:00.
This results in a missing delta for that hour, and if this is the first hour of the date and there is no delta, this results in 100% low carbon energy.

`prev` gets after that the `sum_`  or `0.0` when `sum_` is `none`.

Reply:
This is the data coming directly after the database query:
<img width="1024" alt="CleanShot 2023-10-03 at 22 12 37@2x" src="https://github.com/home-assistant/core/assets/3989428/b10f66e7-4c6b-4070-b5b0-102f479885e6">

And those are the calculated deltas:
<img width="523" alt="CleanShot 2023-10-03 at 22 13 29@2x" src="https://github.com/home-assistant/core/assets/3989428/8a361af1-44cb-4be1-9419-13a3435919c8">

fyi: This is data from my production system, traced with a debugger.

Reply:
We could extend the statistics call to include the previous hour, to calculate the delta for the first hour of the requested period, yes.

Reply:
Ensuring that the calculated delta for the first hour is correct in the new code was my concern, yeah -- I can certainly see the issue with the old code now! Testing in a REPL locally, with the implementation of `_calculate_deltas` from this PR, this is what I saw:

```
>>> _calculate_deltas({0: 10001, 1: 10032, 2: 10045, 3: 10065})
{0: 10001.0, 1: 31, 2: 13, 3: 20}
```

but unless the usage at time `-1` was 0, that first value of 10001 would be incorrect, so I thought in this case the first hour's data would be inflated significantly (e.g. if that first hour was 99% fossil, it'd still show 100 kWh of non-fossil usage, which would likely be more non-fossil usage than actual total usage in that hour)

Reply:
Yea, the only "miscalculation" that could happen in this case is if a sensor that is offline comes back online and reports an initial large amount of energy used.

Reply:
Trying to perform a % or ratio calculation over an arbitrarily small period seems like risky business, and very likely to give misleading results. What if the period is 1 second? I think you can only safely give a result if you are confident that all related sensors have given at least one update since the start of the period, and if the length of the period is at least as large as the longest update interval. That's a complicated constraint so marking the start of the first period as `None` seems like a good compromise, as long as you can prevent calculations being done using `None`.

Reply:
Actually, I don't know how the requested period would be so low without using the websocket directly or abusing/manipulating the frontend.

Anyway, the whole calculation relies on the statistics component, it doesn't collect anything by itself. When data is retrieved, it is summed on an hourly basis for the requested time period:

```python
# Fetch energy + CO2 statistics
statistics = await recorder.get_instance(hass).async_add_executor_job(
    recorder.statistics.statistics_during_period,
    hass,
    start_time,
    end_time,
    statistic_ids,
    "hour",
    {"energy": UnitOfEnergy.KILO_WATT_HOUR},
    {"mean", "sum"},
)
```
So if someone manages to request a 1-second period, we still get an hour of data (which is discarded without this fix).

Also, `_calculate_deltas` can be only called from this energy websocket function.